File: src/test/java/no/fintlabs/flyt/gateway/application/archive/resource/kodeverk/ArkivressursDisplayNameMapperTest.java
package no.fintlabs.flyt.gateway.application.archive.resource.kodeverk;

import no.fint.model.felles.kompleksedatatyper.Identifikator;
import no.fint.model.felles.kompleksedatatyper.Personnavn;
import no.fint.model.resource.Link;
import no.fint.model.resource.administrasjon.personal.PersonalressursResource;
import no.fint.model.resource.arkiv.noark.ArkivressursResource;
import no.fint.model.resource.felles.PersonResource;
import no.fintlabs.cache.FintCache;
import no.fintlabs.cache.exceptions.NoSuchCacheEntryException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.List;
import java.util.Optional;

import static org.assertj.core.api.Assertions.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.*;


@ExtendWith(MockitoExtension.class)
class ArkivressursDisplayNameMapperTest {

    @Mock
    FintCache<String, PersonalressursResource> personalressursResourceCache;

    @Mock
    FintCache<String, PersonResource> personResourceCache;

    ArkivressursDisplayNameMapper arkivressursDisplayNameMapper;

    @BeforeEach
    public void setUp() {
        arkivressursDisplayNameMapper = new ArkivressursDisplayNameMapper(
                personalressursResourceCache,
                personResourceCache
        );
    }

    private ArkivressursResource setupMocksForPersonnavn(Personnavn personnavn) {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of(Link.with("a/b/c/testPersonalressursLink1")));

        PersonalressursResource personalressursResource = mock(PersonalressursResource.class);
        when(personalressursResourceCache.get("a/b/c/testPersonalressursLink1")).thenReturn(personalressursResource);
        when(personalressursResource.getPerson()).thenReturn(List.of(Link.with("a/b/c/testPersonLink1")));

        PersonResource personResource = mock(PersonResource.class);
        when(personResourceCache.get("a/b/c/testPersonLink1")).thenReturn(personResource);

        when(personResource.getNavn()).thenReturn(personnavn);


        return arkivressursResource;
    }

    private ArkivressursResource setupMocksForPersonBrukernavn() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of(Link.with("a/b/c/testPersonalressursLink1")));

        PersonalressursResource personalressursResource = mock(PersonalressursResource.class);
        when(personalressursResourceCache.getOptional("a/b/c/testPersonalressursLink1")).thenReturn(Optional.of(personalressursResource));

        Identifikator identifikator = new Identifikator();
        identifikator.setIdentifikatorverdi("12345");
        when(personalressursResource.getBrukernavn()).thenReturn(identifikator);

        return arkivressursResource;

    }

    @Test
    public void findPersonNavn_givenFirstMiddleAndLastName_shouldReturnFullNameWithSpaceSeparations() {
        Personnavn personnavn = mock(Personnavn.class);
        when(personnavn.getFornavn()).thenReturn("testFornavn");
        when(personnavn.getMellomnavn()).thenReturn("testMellomnavn");
        when(personnavn.getEtternavn()).thenReturn("testEtternavn");

        ArkivressursResource arkivressursResource = setupMocksForPersonnavn(personnavn);

        Optional<String> personNavn = arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);

        assertThat(personNavn).isPresent();
        assertThat(personNavn).contains("testFornavn testMellomnavn testEtternavn");
    }

    @Test
    public void findPersonNavn_givenFirstAndLastName_shouldReturnFullNameWithSpaceSeparations() {
        Personnavn personnavn = mock(Personnavn.class);
        when(personnavn.getFornavn()).thenReturn("testFornavn");
        when(personnavn.getMellomnavn()).thenReturn(null);
        when(personnavn.getEtternavn()).thenReturn("testEtternavn");

        ArkivressursResource arkivressursResource = setupMocksForPersonnavn(personnavn);

        Optional<String> personNavn = arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);

        assertThat(personNavn).isPresent();
        assertThat(personNavn).contains("testFornavn testEtternavn");
    }

    @Test
    public void findPersonNavn_givenNoName_shouldReturnNoName() {
        Personnavn personnavn = mock(Personnavn.class);
        when(personnavn.getFornavn()).thenReturn(null);
        when(personnavn.getMellomnavn()).thenReturn(null);
        when(personnavn.getEtternavn()).thenReturn(null);

        ArkivressursResource arkivressursResource = setupMocksForPersonnavn(personnavn);

        Optional<String> personNavn = arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);

        assertThat(personNavn).isPresent();
        assertThat(personNavn).contains("");
    }

    @Test
    public void findPersonNavn_givenNullName_shouldThrowIllegalStateException() {
        ArkivressursResource arkivressursResource = setupMocksForPersonnavn(null);

        IllegalStateException illegalStateException = assertThrows(
                IllegalStateException.class,
                () -> arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource)
        );

        assertThat(illegalStateException).hasMessage("Person resource contains no name");
    }

    @Test
    public void givenNoMatchingPersonalressursInCache_whenFindPersonNavn_shouldReturnEmpty() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of(Link.with("a/b/c/testPersonalressursLink1")));

        when(personalressursResourceCache.get("a/b/c/testPersonalressursLink1")).thenThrow(
                new NoSuchCacheEntryException("a/b/c/testPersonalressursLink1")
        );

        Optional<String> personNavn = arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);

        assertThat(personNavn).isEmpty();
    }

    @Test
    public void givenNoMatchingPersonInCache_whenFindPersonNavn_shouldReturnEmpty() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of(Link.with("a/b/c/testPersonalressursLink1")));

        PersonalressursResource personalressursResource = mock(PersonalressursResource.class);
        when(personalressursResourceCache.get("a/b/c/testPersonalressursLink1")).thenReturn(personalressursResource);
        when(personalressursResource.getPerson()).thenReturn(List.of(Link.with("a/b/c/testPersonLink1")));

        when(personResourceCache.get("a/b/c/testPersonLink1")).thenThrow(
                new NoSuchCacheEntryException("a/b/c/testPersonLink1")
        );

        Optional<String> personNavn = arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);

        assertThat(personNavn).isEmpty();
    }

    @Test
    public void whenFindPersonNavn_shouldUseFirstLinksToGetFromCache() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of(
                Link.with("a/b/c/testPersonalressursLink1"),
                Link.with("a/b/c/testPersonalressursLink2"),
                Link.with("a/b/c/testPersonalressursLink3")
        ));

        PersonalressursResource personalressursResource = mock(PersonalressursResource.class);
        when(personalressursResourceCache.get("a/b/c/testPersonalressursLink1")).thenReturn(personalressursResource);
        when(personalressursResource.getPerson()).thenReturn(List.of(
                Link.with("a/b/c/testPersonLink1"),
                Link.with("a/b/c/testPersonLink2"),
                Link.with("a/b/c/testPersonLink3")
        ));
        PersonResource personResource = mock(PersonResource.class);
        when(personResourceCache.get("a/b/c/testPersonLink1")).thenReturn(personResource);

        Personnavn personnavn = mock(Personnavn.class);
        when(personnavn.getFornavn()).thenReturn(null);
        when(personnavn.getMellomnavn()).thenReturn(null);
        when(personnavn.getEtternavn()).thenReturn(null);
        when(personResource.getNavn()).thenReturn(personnavn);

        arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);

        verify(personalressursResourceCache, times(1)).get("a/b/c/testPersonalressursLink1");
        verifyNoMoreInteractions(personalressursResourceCache);

        verify(personResourceCache, times(1)).get("a/b/c/testPersonLink1");
        verifyNoMoreInteractions(personResourceCache);
    }

    @Test
    public void givenMatchingPersonalressursInCache_getPersonalressursBrukernavn_shouldReturnIdentifikator() {
        ArkivressursResource arkivressursResource = setupMocksForPersonBrukernavn();

        Optional<String> personBrukernavn = arkivressursDisplayNameMapper.findPersonalressursBrukernavn(arkivressursResource);

        assertThat(personBrukernavn).isPresent();
        assertThat(personBrukernavn).contains("12345");
    }

    @Test
    public void givenNoMatchingPersonalressursInCache_getPersonalressursBrukernavn_shouldReturnLastPartOfPath() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of(Link.with("a/b/c/testPersonalressursLink1")));

        when(personalressursResourceCache.getOptional("a/b/c/testPersonalressursLink1")).thenReturn(Optional.empty());
        Optional<String> personBrukernavn = arkivressursDisplayNameMapper.findPersonalressursBrukernavn(arkivressursResource);

        assertThat(personBrukernavn).isPresent();
        assertThat(personBrukernavn).contains("testPersonalressursLink1");
    }

    @Test
    public void givenNoPersonalressursHrefInArkivressurs_whenFindPersonNavn_shouldReturnOptionalEmpty() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of());
        Optional<String> personNavn = arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource);
        assertThat(personNavn).isEmpty();
    }

    @Test
    public void givenNoPersonalressursHrefInArkivressurs_whenFindPersonalressursBrukernavn_shouldReturnOptionalEmpty() {
        ArkivressursResource arkivressursResource = mock(ArkivressursResource.class);
        when(arkivressursResource.getPersonalressurs()).thenReturn(List.of());
        Optional<String> personalressursBrukernavn = arkivressursDisplayNameMapper.findPersonalressursBrukernavn(arkivressursResource);
        assertThat(personalressursBrukernavn).isEmpty();
    }

}

File: src/test/kotlin/no/novari/flyt/archive/gateway/resource/kodeverk/ArkivressursDisplayNameMapperTest.kt
package no.novari.flyt.archive.gateway.resource.kodeverk

import no.novari.cache.FintCache
import no.novari.cache.exceptions.NoSuchCacheEntryException
import no.novari.fint.model.felles.kompleksedatatyper.Identifikator
import no.novari.fint.model.felles.kompleksedatatyper.Personnavn
import no.novari.fint.model.resource.Link
import no.novari.fint.model.resource.administrasjon.personal.PersonalressursResource
import no.novari.fint.model.resource.arkiv.noark.ArkivressursResource
import no.novari.fint.model.resource.felles.PersonResource
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.Mock
import org.mockito.junit.jupiter.MockitoExtension
import org.mockito.kotlin.mock
import org.mockito.kotlin.times
import org.mockito.kotlin.verify
import org.mockito.kotlin.verifyNoMoreInteractions
import org.mockito.kotlin.whenever
import java.util.Optional

@ExtendWith(MockitoExtension::class)
class ArkivressursDisplayNameMapperTest {
    @Mock
    lateinit var personalressursResourceCache: FintCache<String, PersonalressursResource>

    @Mock
    lateinit var personResourceCache: FintCache<String, PersonResource>

    private lateinit var arkivressursDisplayNameMapper: ArkivressursDisplayNameMapper

    @BeforeEach
    fun setUp() {
        arkivressursDisplayNameMapper = ArkivressursDisplayNameMapper(personalressursResourceCache, personResourceCache)
    }

    @Test
    fun `findPersonNavn given first, middle and last name, returns the full name`() {
        val personnavn: Personnavn = mock()
        whenever(personnavn.fornavn).thenReturn("testFornavn")
        whenever(personnavn.mellomnavn).thenReturn("testMellomnavn")
        whenever(personnavn.etternavn).thenReturn("testEtternavn")

        val arkivressursResource = setupMocksForPersonnavn(personnavn)

        assertThat(arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource))
            .contains("testFornavn testMellomnavn testEtternavn")
    }

    @Test
    fun `findPersonNavn given a null name, throws IllegalStateException`() {
        val arkivressursResource = setupMocksForPersonnavn(null)

        val error =
            assertThrows<IllegalStateException> {
                arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource)
            }

        assertThat(error).hasMessage("Person resource contains no name")
    }

    @Test
    fun `findPersonNavn given no matching personalressurs in cache, returns empty`() {
        val arkivressursResource: ArkivressursResource = mock()
        whenever(arkivressursResource.personalressurs).thenReturn(listOf(Link.with("a/b/c/testPersonalressursLink1")))
        whenever(personalressursResourceCache.get("a/b/c/testPersonalressursLink1"))
            .thenThrow(NoSuchCacheEntryException("a/b/c/testPersonalressursLink1"))

        assertThat(arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource)).isEmpty
    }

    @Test
    fun `findPersonNavn uses the first links to look up in the cache`() {
        val arkivressursResource: ArkivressursResource = mock()
        whenever(arkivressursResource.personalressurs)
            .thenReturn(
                listOf(
                    Link.with("a/b/c/testPersonalressursLink1"),
                    Link.with("a/b/c/testPersonalressursLink2"),
                ),
            )

        val personalressursResource: PersonalressursResource = mock()
        whenever(
            this.personalressursResourceCache.get("a/b/c/testPersonalressursLink1"),
        ).thenReturn(personalressursResource)
        whenever(
            personalressursResource.person,
        ).thenReturn(listOf(Link.with("a/b/c/testPersonLink1"), Link.with("a/b/c/testPersonLink2")))
        val personResource: PersonResource = mock()
        whenever(personResourceCache.get("a/b/c/testPersonLink1")).thenReturn(personResource)
        val personnavn: Personnavn = mock()
        whenever(personResource.navn).thenReturn(personnavn)

        arkivressursDisplayNameMapper.findPersonNavn(arkivressursResource)

        verify(personalressursResourceCache, times(1)).get("a/b/c/testPersonalressursLink1")
        verifyNoMoreInteractions(personalressursResourceCache)
        verify(personResourceCache, times(1)).get("a/b/c/testPersonLink1")
        verifyNoMoreInteractions(personResourceCache)
    }

    @Test
    fun `findPersonalressursBrukernavn given a matching personalressurs in the cache, returns the identifikator`() {
        val arkivressursResource = setupMocksForPersonBrukernavn()

        assertThat(arkivressursDisplayNameMapper.findPersonalressursBrukernavn(arkivressursResource)).contains("12345")
    }

    private fun setupMocksForPersonnavn(personnavn: Personnavn?): ArkivressursResource {
        val arkivressursResource: ArkivressursResource = mock()
        whenever(arkivressursResource.personalressurs).thenReturn(listOf(Link.with("a/b/c/testPersonalressursLink1")))

        val personalressursResource: PersonalressursResource = mock()
        whenever(personalressursResourceCache.get("a/b/c/testPersonalressursLink1")).thenReturn(personalressursResource)
        whenever(personalressursResource.person).thenReturn(listOf(Link.with("a/b/c/testPersonLink1")))

        val personResource: PersonResource = mock()
        whenever(personResourceCache.get("a/b/c/testPersonLink1")).thenReturn(personResource)
        whenever(personResource.navn).thenReturn(personnavn)

        return arkivressursResource
    }

    private fun setupMocksForPersonBrukernavn(): ArkivressursResource {
        val arkivressursResource: ArkivressursResource = mock()
        whenever(arkivressursResource.personalressurs).thenReturn(listOf(Link.with("a/b/c/testPersonalressursLink1")))

        val personalressursResource: PersonalressursResource = mock()
        whenever(personalressursResourceCache.getOptional("a/b/c/testPersonalressursLink1"))
            .thenReturn(Optional.of(personalressursResource))

        val identifikator = Identifikator()
        identifikator.identifikatorverdi = "12345"
        whenever(personalressursResource.brukernavn).thenReturn(identifikator)

        return arkivressursResource
    }
}
